assert.go

```diff
@@ -27,33 +27,63 @@ type Tester interface {
 	Fatalf(string, ...interface{})
 }
 
+// frameWrapper fulfills the Tester interface and is a simple wrapper around another Tester (which
+// itself could be a frameWrapper). frameWrapper doesn't add any specific functionality, but it is
+// a marker used by the callerStr function, which unwraps successive layers of frameWrappers to
+// understand how far back in the call stack it needs to go to display the correct source of a
+// failed assertion.
+type frameWrapper struct {
+	t Tester
+}
+
+func (f frameWrapper) Fatalf(fmtStr string, vals ...interface{}) {
+	f.t.Fatalf(fmtStr, vals...)
+}
+
+// WithFrameWrapper returns the original Tester, wrapped by a frameWrapper that also implements
+// Tester. Layers of such wrappers track the number of frames on the call stack to backtrack when
+// identifying the source of an assertion failure.
+func WithFrameWrapper(t Tester) Tester {
+	return &frameWrapper{t: t}
+}
+
 // callerStr returns a string representation of the code numFrames stack
 // frames above the code that called callerStr
-func callerStr(numFrames int) string {
-	_, file, line, _ := runtime.Caller(1 + numFrames)
+func callerStr(t Tester) string {
+	localT := t
+	numFrames := 1
+	ok := true
+	for ok {
+		var fw *frameWrapper
+		if fw, ok = localT.(*frameWrapper); ok {
+			numFrames++
```

Comment:
@krancour let me suggest an addition:

- Make `frameWrapper` exported
- Make the number of frames to traverse configurable on the newly exported `FrameWrapper`
- Increment by that number of frames here